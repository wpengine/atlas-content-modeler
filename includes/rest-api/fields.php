<?php
/**
 * Field helpers used in REST API callbacks.
 *
 * @package AtlasContentModeler
 */

namespace WPE\AtlasContentModeler\REST_API\Fields;

use function WPE\AtlasContentModeler\ContentRegistration\get_registered_content_types;
use function WPE\AtlasContentModeler\ContentRegistration\update_registered_content_types;
use WPE\AtlasContentModeler\ContentConnect\Plugin as ContentConnect;

/**
 * Shapes the field arguments array.
 *
 * @param array $args The field arguments.
 *
 * @return array
 */
function shape_field_args( array $args ): array {
	$defaults = [
		'show_in_rest'    => true,
		'show_in_graphql' => true,
	];

	$merged = array_merge( $defaults, $args );

	unset( $merged['_locale'] ); // Sent by wp.apiFetch but not needed.
	unset( $merged['model'] ); // The field is stored in the fields property of its model.

	return $merged;
}

/**
 * Deletes relationship fields that refer to the `$slug` model from all models.
 *
 * @param string $slug The reference model to delete relationship fields for.
 */
function cleanup_detached_relationship_fields( string $slug ) {
	$models = get_registered_content_types();

	foreach ( $models as $model_id => $model ) {
		foreach ( $model['fields'] as $field_id => $field ) {
			if ( $field['type'] === 'relationship' && $field['reference'] === $slug ) {
				unset( $models[ $model_id ]['fields'][ $field_id ] );
			}
		}
	}

	update_registered_content_types( $models );
}

/**
 * Deletes rows in the post-to-post table whose id1 value refers to a post that
 * has the `$slug` post type.
 *
 * This depends on model entries for `$slug` existing in the database. At the
 * moment we do not delete entries when a model is deleted. If that changes,
 * this cleanup function must run before entries are removed from the database.
 *
 * TODO: When we implement backreferences, clean those here too.
 *
 * TODO: This could be moved to the acm-content-connect package when a hook
 * exists for deleting ACM models.
 *
 * @param string $slug The reference model to delete relationship entries for.
 */
function cleanup_detached_relationship_references( string $slug ) {
	global $wpdb;

	$table        = ContentConnect::instance()->get_table( 'p2p' );
	$post_to_post = $table->get_table_name();

	/**
	 * PHPCS is disabled to prevent warnings about the unescaped `$post_to_post`
	 * table name, which is derived from an unfilterable string literal in
	 * WPE\AtlasContentModeler\ContentConnect\Tables\PostToPost\get_table_name.
	 *
	 * Needed feature to fix this: https://core.trac.wordpress.org/ticket/52506.
	 *
	 * phpcs:disable
	 */
	$wpdb->query(
		$wpdb->prepare(
			"
			DELETE `{$post_to_post}`
			FROM `{$post_to_post}`
			INNER JOIN {$wpdb->posts} ON `{$post_to_post}`.id1 = {$wpdb->posts}.ID
			WHERE {$wpdb->posts}.post_type = %s;
			",
			$slug
		)
	);
	// phpcs:enable
}

/**
 * Checks if a duplicate field identifier (slug) exists in the content model.
 *
 * @param string $slug  The current field slug.
 * @param string $id    The current field id.
 * @param string $model_id The id of the content model to check for duplicate slugs.
 * @return bool
 */
function content_model_field_exists( string $slug, string $id, string $model_id ): bool {
	$models = get_registered_content_types();
	$model  = $models[ $model_id ] ?? [];

	if ( ! isset( $model['fields'] ) ) {
		return false;
	}

	foreach ( $model['fields'] as $field ) {
		if ( $field['id'] === $id ) {
			continue;
		}

		if ( $field['slug'] === $slug ) {
			return true;
		}
	}

	return false;
}

/**
 * Checks if a relationship field has the same reverse slug as other
 * relationship fields in the same model.
 *
 * @param array $candidate_field The field to check other relationship fields
 *                               for a reverse slug collision with.
 * @return bool True if another relationship field has the same reverse slug
 *              and relates to the same model.
 */
function content_model_reverse_slug_exists( $candidate_field ) {
	$models = get_registered_content_types();
	$fields = $models[ $candidate_field['model'] ]['fields'] ?? [];

	foreach ( $fields as $current_field ) {
		$reverse_enabled = $current_field['enableReverse'] ?? false;

		if (
			$current_field['type'] !== 'relationship' ||
			$current_field['id'] === $candidate_field['id'] ||
			! $reverse_enabled
		) {
			continue;
		}

		if (
			( $current_field['reverseSlug'] ?? '' ) === ( $candidate_field['reverseSlug'] ?? false ) &&
			( $current_field['reference'] ?? '' ) === ( $candidate_field['reference'] ?? false )
		) {
			return true;
		}
	}

	return false;
}

/**
 * Checks if a duplicate model identifier (name) exists in the multiple option field.
 *
 * @param array  $names  The available field choice names.
 * @param string $current_choice  The currently checked field choice name.
 * @param int    $current_index The content index for the current choice being validated.
 * @return bool
 */
function content_model_multi_option_exists( array $names, string $current_choice, int $current_index ): bool {
	if ( ! isset( $names ) ) {
		return false;
	}
	if ( $names ) {
		if ( $names[ $current_index ] ) {
			unset( $names[ $current_index ] );
		}

		foreach ( $names as $choice ) {
			if ( $choice['name'] === $current_choice ) {
				return true;
			}
		}
	}
	return false;
}

/**
 * Checks if a duplicate model identifier (slug) exists in the multiple option field.
 *
 * @param array  $slugs  The available field slug names.
 * @param string $current_choice  The currently checked field choice name.
 * @param int    $current_index The content index for the current choice being validated.
 * @return bool
 */
function content_model_multi_option_slug_exists( array $slugs, string $current_choice, int $current_index ): bool {
	if ( ! isset( $slugs ) ) {
		return false;
	}
	if ( $slugs ) {
		if ( $slugs[ $current_index ] ) {
			unset( $slugs[ $current_index ] );
		}

		foreach ( $slugs as $choice ) {
			if ( $choice['slug'] === $current_choice ) {
				return true;
			}
		}
	}
	return false;
}
