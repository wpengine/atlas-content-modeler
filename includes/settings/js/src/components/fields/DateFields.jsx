/**
 * Additional form fields for the Date field type.
 */
import React, { useState } from "react";
import { __ } from "@wordpress/i18n";

const DateFields = ({ register, data, editing, fieldId }) => {
	const [showRepeatableDate, setShowRepeatableDate] = useState(
		data?.isRepeatableDate === true
	);
	return (
		<>
			{data && (
				<div className={"field"}>
					<legend>
						{__("Repeatable Date Field", "atlas-content-modeler")}
					</legend>
					<input
						// Each name should be unique, so for this field we are using isRepeatableDate.
						name="isRepeatableDate"
						type="checkbox"
						id={`is-repeatable-date-${fieldId}`}
						ref={register}
						value={showRepeatableDate}
						onChange={() =>
							setShowRepeatableDate(!showRepeatableDate)
						}
						disabled={editing}
					/>
					<label
						htmlFor={`is-repeatable-date-${fieldId}`}
						className="checkbox is-repeatable"
					>
						{__(
							"Make this field repeatable",
							"atlas-content-modeler"
						)}
					</label>
				</div>
			)}
		</>
	);
};

export default DateFields;
