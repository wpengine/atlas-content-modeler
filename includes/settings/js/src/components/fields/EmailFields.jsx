/**
 * Additional form fields for the Email field type.
 */
import React, { useContext, useState } from "react";
import { getTitleFieldId } from "../../queries";
import { ModelsContext } from "../../ModelsContext";
import { useLocationSearch } from "../../utils";
import { __ } from "@wordpress/i18n";

const EmailFields = ({ register, data, editing, fieldId }) => {
	const { models } = useContext(ModelsContext);
	const query = useLocationSearch();
	const currentModel = query.get("id");
	const fields = models[currentModel]?.fields;
	const titleFieldId = getTitleFieldId(fields);
	const showTitleField = !titleFieldId || titleFieldId === fieldId;
	const [showTitle, setShowTitle] = useState(data?.isTitle);
	const [showRepeatableEmail, setShowRepeatableEmail] = useState(
		data?.isRepeatableEmail
	);

	return (
		<>
			{data && (
				<div className={"field"}>
					<legend>
						{__("Repeatable Field", "atlas-content-modeler")}
					</legend>
					<input
						name="isRepeatableEmail"
						type="checkbox"
						id={`is-repeatable-email-${fieldId}`}
						ref={register}
						defaultChecked={showRepeatableEmail}
						onChange={() =>
							setShowRepeatableEmail(!showRepeatableEmail)
						}
						disabled={editing || showTitle}
					/>
					<label
						htmlFor={`is-repeatable-email-${fieldId}`}
						className="checkbox is-repeatable"
					>
						{__(
							"Make this field repeatable",
							"atlas-content-modeler"
						)}
					</label>
				</div>
			)}
			{showTitleField && (
				<div className="field">
					<legend>Title Field</legend>
					<input
						name="isTitle"
						type="checkbox"
						id={`is-title-${fieldId}`}
						ref={register}
						onChange={() => setShowTitle(!showTitle)}
						defaultChecked={data?.isTitle}
						disabled={!!titleFieldId || showRepeatableEmail}
					/>
					<label
						htmlFor={`is-title-${fieldId}`}
						className="checkbox is-title"
					>
						{__(
							"Use this field as the entry title",
							"atlas-content-modeler"
						)}
					</label>
				</div>
			)}
		</>
	);
};

export default EmailFields;
