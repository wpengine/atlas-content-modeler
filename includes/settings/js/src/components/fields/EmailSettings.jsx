/** @jsx jsx */
import { jsx, css } from "@emotion/react";
import React from "react";
import Icon from "../../../../../components/icons";
import { __ } from "@wordpress/i18n";

const emailDomainTextArea = css`
	.atlas-content-modeler-admin-page & {
		max-width: none;
		min-height: 55px;
	}
`;

const emailConstraintInput = css`
	.atlas-content-modeler-admin-page & {
		max-width: 150px;
	}
`;

const EmailSettings = ({
	errors,
	storedData,
	setValue,
	getValues,
	trigger,
	watch,
}) => {
	watch(["minRepeatable", "maxRepeatable", "exactRepeatable"]);
	const isRepeatable = getValues("isRepeatableEmail");
	const isMinMaxDisabled = !!getValues("exactRepeatable");
	const isExactDisabled =
		getValues("maxRepeatable") || !!getValues("minRepeatable");

	return (
		<>
			<div className="d-flex flex-column d-sm-flex flex-sm-row">
				<div className="w-100">
					<div
						className={`${
							errors.allowedDomains ? "field has-error" : "field"
						} w-100`}
					>
						<label htmlFor="allowedDomains">
							{__(
								"Specify Allowed Domains",
								"atlas-content-modeler"
							)}
						</label>
						<p>
							{__(
								`Comma-separated list of allowed domains, wildcards allowed. e.g. "*.edu"`,
								"atlas-content-modeler"
							)}
						</p>
						<textarea
							id="allowedDomains"
							name="allowedDomains"
							css={emailDomainTextArea}
							className="w-100"
							rows="1"
							aria-invalid={
								errors.allowedDomains ? "true" : "false"
							}
							placeholder={__(
								"*.edu, gmail.com, wpengine.com, *.flywheel.com",
								"atlas-content-modeler"
							)}
							onChange={async (e) => {
								setValue("allowedDomains", e.target.value, {
									shouldValidate: true,
								});
							}}
							defaultValue={String(
								getValues("allowedDomains") ??
									storedData?.allowedDomains
							)}
						/>
						<p className="field-messages">
							{errors.allowedDomains &&
								errors.allowedDomains.type ===
									"formattedCorrectly" && (
									<span className="error">
										<Icon type="error" />
										<span
											role="alert"
											className="text-start"
										>
											{__(
												"Must be a comma-separated list of domains and subdomains.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>
				</div>
			</div>

			{isRepeatable && (
				<div className="d-flex flex-column d-sm-flex flex-sm-row justify-content-between">
					<div
						className={
							errors.maxRepeatable ? "field has-error" : "field"
						}
					>
						<label htmlFor="minRepeatable">
							{__("Minimum", "atlas-content-modeler")}
						</label>
						<br />
						<input
							id="minRepeatable"
							name="minRepeatable"
							type="number"
							css={emailConstraintInput}
							aria-invalid={
								errors.minRepeatable ? "true" : "false"
							}
							defaultValue={String(
								getValues("minRepeatable") ??
									storedData?.minRepeatable
							)}
							onChange={async (e) => {
								setValue("minRepeatable", e.target.value, {
									shouldValidate: true,
								});
								// Validate maxRepeatable in case minRepeatable is now bigger.
								await trigger("maxRepeatable");
							}}
							disabled={isMinMaxDisabled}
						/>
						<p className="field-messages">
							{errors.minRepeatable &&
								errors.minRepeatable.type === "min" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"The minimum value is 0.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>

					<div
						className={
							errors.maxRepeatable ? "field has-error" : "field"
						}
					>
						<label htmlFor="maxRepeatable">
							{__("Maximum", "atlas-content-modeler")}
						</label>
						<br />
						<input
							id="maxRepeatable"
							name="maxRepeatable"
							type="number"
							css={emailConstraintInput}
							aria-invalid={
								errors.maxRepeatable ? "true" : "false"
							}
							defaultValue={String(
								getValues("maxRepeatable") ??
									storedData?.maxRepeatable
							)}
							onChange={(e) => {
								setValue("maxRepeatable", e.target.value, {
									shouldValidate: true,
								});
							}}
							disabled={isMinMaxDisabled}
						/>
						<p className="field-messages">
							{errors.maxRepeatable &&
								errors.maxRepeatable.type === "min" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"The minimum value is 1.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
							{errors.maxRepeatable &&
								errors.maxRepeatable.type === "maxBelowMin" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"Max must be more than min.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>

					<div
						className={
							errors.exactRepeatable ? "field has-error" : "field"
						}
					>
						<label htmlFor="exactRepeatable">
							{__("Exact", "atlas-content-modeler")}
						</label>
						<br />
						<input
							id="exactRepeatable"
							name="exactRepeatable"
							type="number"
							css={emailConstraintInput}
							aria-invalid={
								errors.exactRepeatable ? "true" : "false"
							}
							defaultValue={String(
								getValues("exactRepeatable") ??
									storedData?.exactRepeatable
							)}
							onChange={(e) => {
								setValue("exactRepeatable", e.target.value, {
									shouldValidate: true,
								});
							}}
							disabled={isExactDisabled}
						/>
						<p className="field-messages">
							{errors.exactRepeatable &&
								errors.exactRepeatable.type === "min" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"The minimum value is 1.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>
				</div>
			)}
		</>
	);
};

export default EmailSettings;
