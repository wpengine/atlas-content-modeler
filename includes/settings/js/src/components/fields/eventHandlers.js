import { POSITION_GAP } from "../../queries";
import { toast } from "react-toastify";
import { showError } from "../../toasts";
import { sprintf, __ } from "@wordpress/i18n";

const { apiFetch } = wp;

/**
 * Reorders `list` by moving the item at `startIndex` to `endIndex`.
 *
 * @param list
 * @param startIndex
 * @param endIndex
 * @return {Array}
 * @example
 * ```
 * reorder(["a", "b", "c"], 0, 1)); // ["b", "a", "c"]
 * reorder(["a", "b", "c"], 0, 2)); // ["b", "c", "a"]
 * ```
 */
const reorder = (list, startIndex, endIndex) => {
	const result = Array.from(list);
	const [removed] = result.splice(startIndex, 1);
	result.splice(endIndex, 0, removed);

	return result;
};

/**
 * Updates field positions in the WordPress database.
 *
 * @return {Promise<void>}
 */
const updatePositions = async (model, idsAndPositions) => {
	await apiFetch({
		path: `/wpe/atlas/content-model-fields/${model}`,
		method: "PATCH",
		_wpnonce: wpApiSettings.nonce,
		data: { fields: idsAndPositions },
	});
};

/**
 * Updates the model store and WordPress database when fields are reordered.
 */
export function onDragEnd(result, fields, model, dispatch, models) {
	const { destination, source } = result;

	if (!destination) {
		return;
	}

	if (
		destination.droppableId === source.droppableId &&
		destination.index === source.index
	) {
		return;
	}

	// Store original field order to revert if the position update fails.
	const idsAndOldPositions = fields.reduce((result, id) => {
		result[id] = { position: models[model]["fields"][id]?.position };
		return result;
	}, {});

	const newOrder = reorder(
		fields,
		result.source.index,
		result.destination.index
	);

	let position = 0;
	const idsAndNewPositions = newOrder.reduce((result, id) => {
		result[id] = { position };
		position += POSITION_GAP;
		return result;
	}, {});

	// Optimistically update the client-side model store so the new field order is rendered immediately.
	dispatch({
		type: "reorderFields",
		positions: idsAndNewPositions,
		model: model,
	});

	updatePositions(model, idsAndNewPositions)
		.then((res) => {
			toast.dismiss("error");
		})
		.catch((err) => {
			showError(
				sprintf(
					__(
						"Error saving field order: %s. Close this and try again?",
						"atlas-content-modeler"
					),
					err?.message ||
						__("Error unknown.", "atlas-content-modeler")
				)
			);

			// Revert local field order so the state is accurate and the user can retry.
			dispatch({
				type: "reorderFields",
				positions: idsAndOldPositions,
				model: model,
			});
		});
}
