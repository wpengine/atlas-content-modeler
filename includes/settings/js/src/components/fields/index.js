import Field from './Field';

export { Field };
