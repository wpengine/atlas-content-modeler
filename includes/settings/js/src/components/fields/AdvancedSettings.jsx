/**
 * Additional field options that appear in the Advanced Settings modal.
 */
import React from "react";
import Icon from "../../../../../components/icons";
import { __ } from "@wordpress/i18n";
import EmailSettings from "./EmailSettings";

const MediaSettings = ({
	errors,
	storedData,
	setValue,
	getValues,
	trigger,
}) => {
	return (
		<>
			<h3>{__("Allowed File Types", "atlas-content-modeler")}</h3>
			<p>
				{__(
					"Define what file types are allowed to be uploaded.",
					"atlas-content-modeler"
				)}
			</p>

			<p className="mb-4">
				<a
					href="https://wpengine.com/support/mime-types-wordpress/"
					target="_blank"
					rel="noreferrer"
				>
					{__("WordPress Allowed Types", "atlas-content-modeler")}
				</a>
			</p>

			<div className="d-flex flex-column d-sm-flex flex-sm-row">
				<div className="w-100">
					<div
						className={`${
							errors.allowedTypes ? "field has-error" : "field"
						} w-100`}
					>
						<label htmlFor="allowedTypes">
							{__("File Extensions", "atlas-content-modeler")}
						</label>
						<br />
						<textarea
							className="w-100"
							aria-invalid={
								errors.allowedTypes ? "true" : "false"
							}
							placeholder={__(
								"jpeg,jpg,png,tiff,tif,pdf",
								"atlas-content-modeler"
							)}
							id="allowedTypes"
							name="allowedTypes"
							onChange={async (e) => {
								setValue("allowedTypes", e.target.value, {
									shouldValidate: true,
								});
							}}
							defaultValue={String(
								getValues("allowedTypes") ??
									storedData?.allowedTypes
							)}
						/>
						<p className="field-messages">
							{errors.allowedTypes &&
								errors.allowedTypes.type ===
									"formattedCorrectly" && (
									<span className="error">
										<Icon type="error" />
										<span
											role="alert"
											className="text-start"
										>
											{__(
												"Must be a comma-separated list of file extensions without periods.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>
				</div>
			</div>
		</>
	);
};

const NumberSettings = ({
	errors,
	storedData,
	setValue,
	getValues,
	trigger,
}) => {
	function getDefaultNumberValue(value, storedValue) {
		return value ?? storedValue;
	}
	return (
		<>
			<h3>{__("Specific Number Range", "atlas-content-modeler")}</h3>
			<p className="mb-4">
				{__(
					"Define your range here. Only numeric characters can be used.",
					"atlas-content-modeler"
				)}
			</p>

			<div className="d-flex flex-column d-sm-flex flex-sm-row">
				<div>
					<div
						className={`${
							errors.minValue ? "field has-error" : "field"
						} me-sm-5`}
					>
						<label htmlFor="minValue">
							{__("Min Value", "atlas-content-modeler")}
						</label>
						<br />
						<input
							aria-invalid={errors.minValue ? "true" : "false"}
							type="number"
							step="0.1"
							id="minValue"
							name="minValue"
							onChange={async (e) => {
								setValue("minValue", e.target.value, {
									shouldValidate: true,
								});
								// Validate maxValue in case minValue is now bigger.
								await trigger("maxValue");
								await trigger("step");
							}}
							defaultValue={getDefaultNumberValue(
								getValues("minValue"),
								storedData?.minValue
							)}
						/>
						<p className="field-messages">
							{errors.minValue && errors.minValue.type === "min" && (
								<span className="error">
									<Icon type="error" />
									<span role="alert">
										{__(
											"The minimum value is 0.",
											"atlas-content-modeler"
										)}
									</span>
								</span>
							)}
							{errors.minValue &&
								errors.minValue.type ===
									"isAppropriateType" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"The value must be an integer.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>

					<div
						className={`${
							errors.maxValue ? "field has-error" : "field"
						} me-sm-5`}
					>
						<label htmlFor="maxValue">
							{__("Max Value", "atlas-content-modeler")}
						</label>
						<br />
						<input
							aria-invalid={errors.maxValue ? "true" : "false"}
							type="number"
							step="0.1"
							id="maxValue"
							name="maxValue"
							onChange={async (e) => {
								setValue("maxValue", e.target.value, {
									shouldValidate: true,
								});
								await trigger("step");
							}}
							defaultValue={getDefaultNumberValue(
								getValues("maxValue"),
								storedData?.maxValue
							)}
						/>
						<p className="field-messages">
							{errors.maxValue && errors.maxValue.type === "min" && (
								<span className="error">
									<Icon type="error" />
									<span role="alert">
										{__(
											"The minimum value is 0.",
											"atlas-content-modeler"
										)}
									</span>
								</span>
							)}
							{errors.maxValue &&
								errors.maxValue.type === "maxBelowMin" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"Max must be more than min.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
							{errors.maxValue &&
								errors.maxValue.type ===
									"isAppropriateType" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"The value must be an integer.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>
				</div>

				<div className={errors.step ? "field has-error" : "field"}>
					<label htmlFor="step">
						{__("Step", "atlas-content-modeler")}
					</label>
					<br />
					<input
						aria-invalid={errors.step ? "true" : "false"}
						type="number"
						step="0.1"
						id="step"
						name="step"
						onChange={(e) => {
							setValue("step", e.target.value, {
								shouldValidate: true,
							});
						}}
						defaultValue={getDefaultNumberValue(
							getValues("step"),
							storedData?.step
						)}
					/>
					<p className="field-messages">
						{errors.step && errors.step.type === "min" && (
							<span className="error">
								<Icon type="error" />
								<span role="alert">
									{__(
										"The minimum value is 0.",
										"atlas-content-modeler"
									)}
								</span>
							</span>
						)}
						{errors.step && errors.step.type === "maxBelowStep" && (
							<span className="error">
								<Icon type="error" />
								<span role="alert">
									{__(
										"Step must be lower than max.",
										"atlas-content-modeler"
									)}
								</span>
							</span>
						)}
						{errors.step &&
							errors.step.type ===
								"minAndStepEqualOrLessThanMax" && (
								<span className="error">
									<Icon type="error" />
									<span role="alert">
										{__(
											"Min plus Step can't be larger than Max.",
											"atlas-content-modeler"
										)}
									</span>
								</span>
							)}
						{errors.step &&
							errors.step.type === "isAppropriateType" && (
								<span className="error">
									<Icon type="error" />
									<span role="alert">
										{__(
											"The value must be an integer.",
											"atlas-content-modeler"
										)}
									</span>
								</span>
							)}
					</p>
				</div>
			</div>
		</>
	);
};

const TextSettings = ({
	errors,
	field,
	storedData,
	setValue,
	getValues,
	trigger,
}) => {
	const isRepeatable = getValues("isRepeatable");
	return (
		<>
			<h3>{__("Character Limit", "atlas-content-modeler")}</h3>
			<p className="mb-4">
				{__(
					"Set a minimum and/or maximum character count for this text field.",
					"atlas-content-modeler"
				)}
			</p>

			<div className="d-flex flex-column d-sm-flex flex-sm-row">
				<div
					className={`${
						errors.minChars ? "field has-error" : "field"
					} me-sm-5`}
				>
					<label htmlFor="minChars">
						{__("Minimum Character Limit", "atlas-content-modeler")}
					</label>
					<br />
					<input
						aria-invalid={errors.minChars ? "true" : "false"}
						type="number"
						id="minChars"
						name="minChars"
						onChange={async (e) => {
							setValue("minChars", e.target.value, {
								shouldValidate: true,
							});
							// Validate maxChars in case minChars is now bigger.
							await trigger("maxChars");
						}}
						defaultValue={String(
							getValues("minChars") ?? storedData?.minChars
						)}
					/>
					<p className="field-messages">
						{errors.minChars && errors.minChars.type === "min" && (
							<span className="error">
								<Icon type="error" />
								<span role="alert">
									{__(
										"The minimum value is 0.",
										"atlas-content-modeler"
									)}
								</span>
							</span>
						)}
					</p>
				</div>

				<div className={errors.maxChars ? "field has-error" : "field"}>
					<label htmlFor="maxChars">
						{__("Maximum Character Limit", "atlas-content-modeler")}
					</label>
					<br />
					<input
						aria-invalid={errors.maxChars ? "true" : "false"}
						type="number"
						id="maxChars"
						name="maxChars"
						onChange={(e) => {
							setValue("maxChars", e.target.value, {
								shouldValidate: true,
							});
						}}
						defaultValue={String(
							getValues("maxChars") ?? storedData?.maxChars
						)}
					/>
					<p className="field-messages">
						{errors.maxChars && errors.maxChars.type === "min" && (
							<span className="error">
								<Icon type="error" />
								<span role="alert">
									{__(
										"The minimum value is 1.",
										"atlas-content-modeler"
									)}
								</span>
							</span>
						)}
						{errors.maxChars &&
							errors.maxChars.type === "maxBelowMin" && (
								<span className="error">
									<Icon type="error" />
									<span role="alert">
										{__(
											"Max must be more than min.",
											"atlas-content-modeler"
										)}
									</span>
								</span>
							)}
					</p>
				</div>
			</div>
			{isRepeatable && (
				<div className="d-flex flex-column d-sm-flex flex-sm-row">
					<div
						className={`${
							errors.minRepeatable ? "field has-error" : "field"
						} me-sm-5`}
					>
						<label htmlFor="minRepeatable">
							{__(
								"Minimum Repeatable Limit",
								"atlas-content-modeler"
							)}
						</label>
						<br />
						<input
							aria-invalid={
								errors.minRepeatable ? "true" : "false"
							}
							type="number"
							id="minRepeatable"
							name="minRepeatable"
							onChange={async (e) => {
								setValue("minRepeatable", e.target.value, {
									shouldValidate: true,
								});
								// Validate maxRepeatable in case minRepeatable is now bigger.
								await trigger("maxRepeatable");
							}}
							defaultValue={String(
								getValues("minRepeatable") ??
									storedData?.minRepeatable
							)}
						/>
						<p className="field-messages">
							{errors.minRepeatable &&
								errors.minRepeatable.type === "min" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"The minimum value is 0.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>

					<div
						className={
							errors.maxRepeatable ? "field has-error" : "field"
						}
					>
						<label htmlFor="maxRepeatable">
							{__(
								"Maximum Repeatable Limit",
								"atlas-content-modeler"
							)}
						</label>
						<br />
						<input
							aria-invalid={
								errors.maxRepeatable ? "true" : "false"
							}
							type="number"
							id="maxRepeatable"
							name="maxRepeatable"
							onChange={(e) => {
								setValue("maxRepeatable", e.target.value, {
									shouldValidate: true,
								});
							}}
							defaultValue={String(
								getValues("maxRepeatable") ??
									storedData?.maxRepeatable
							)}
						/>
						<p className="field-messages">
							{errors.maxRepeatable &&
								errors.maxRepeatable.type === "min" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"The minimum value is 1.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
							{errors.maxRepeatable &&
								errors.maxRepeatable.type === "maxBelowMin" && (
									<span className="error">
										<Icon type="error" />
										<span role="alert">
											{__(
												"Max must be more than min.",
												"atlas-content-modeler"
											)}
										</span>
									</span>
								)}
						</p>
					</div>
				</div>
			)}
		</>
	);
};

export { MediaSettings, TextSettings, NumberSettings, EmailSettings };
