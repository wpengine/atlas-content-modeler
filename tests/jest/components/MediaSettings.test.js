import React from "react";
import { MediaSettings } from "../../../includes/settings/js/src/components/fields/AdvancedSettings";
import ShallowRenderer from "react-test-renderer/shallow";

describe("MediaSettings tests", () => {
	const allowed = "jpeg,jpg,pdf";
	const mock = {
		errors: {},
		getValues: jest.fn(() => {
			return allowed;
		}),
	};

	const renderer = new ShallowRenderer();
	renderer.render(<MediaSettings {...mock} />);

	const tree = renderer.getRenderOutput();

	it("Renders a matching snapshot", () => {
		expect(tree).toMatchSnapshot();
	});
});
