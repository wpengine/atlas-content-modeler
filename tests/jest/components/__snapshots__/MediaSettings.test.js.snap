// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaSettings tests Renders a matching snapshot 1`] = `
<React.Fragment>
  <h3>
    Allowed File Types
  </h3>
  <p>
    Define what file types are allowed to be uploaded.
  </p>
  <p
    className="mb-4"
  >
    <a
      href="https://wpengine.com/support/mime-types-wordpress/"
      rel="noreferrer"
      target="_blank"
    >
      WordPress Allowed Types
    </a>
  </p>
  <div
    className="d-flex flex-column d-sm-flex flex-sm-row"
  >
    <div
      className="w-100"
    >
      <div
        className="field w-100"
      >
        <label
          htmlFor="allowedTypes"
        >
          File Extensions
        </label>
        <br />
        <textarea
          aria-invalid="false"
          className="w-100"
          defaultValue="jpeg,jpg,pdf"
          id="allowedTypes"
          name="allowedTypes"
          onChange={[Function]}
          placeholder="jpeg,jpg,png,tiff,tif,pdf"
        />
        <p
          className="field-messages"
        />
      </div>
    </div>
  </div>
</React.Fragment>
`;
